### PR TITLE
JS-329 Upgrade TypeScript to 5.6.2

### PR DESCRIPTION
**Pull request by saberduck**
Behavior wrt. project references has changed. Previously, if a project overlapped with the referenced project, it would include the files of the referenced project regardless. Now, the files from the referenced project are not included. This is better for us since it avoids analyzing the files duplicitly.

### CODE REVIEW

**Comment by pmaieref** [comment]
Hi, I was about to open a thread in the forum regarding the TypeScript update, but as I see this here now, I thought I might just ask here. If it is too off-topic, please tell me and I'll move it to the community forum.

I am wondering if there are any guidelines, processes, documentation on how TypeScript versions are handled within SonarQube / SonarJS. The documentation does not provide any insights into:
- Can my local project use a higher / lower TS version than SQ/SonarJS
  - Considering this PR makes it into SQ 10.7, do I need my project on TS 5.6 as well to work in SQ 10.7?)
- What is the release cadence of TS updates in SonarJS? I see that 5.5 was skipped and you update directly to TS 5.6. Considering TS 5.6 is now 4 days old, this update is "early" (thanks for that!), but "support" for TS 5.5 is quite late.
- How do certain TS compiler flags play into this? Does SQ need to explicitly support / drop support for them?

The project I am working on, strives to use current / up-to-date versions of Angular, but in the past SQ's TS version was a blocking issue that prevented updating Angular as they raise the minimum versions of TS every 6 months with every major update. We ran into (compilation / execution) issues due to different TS versions and had to downgrade again. Now we don't even try if it would maybe work and just wait it out.

Thanks in advance.

**Comment by gabriel-vivas-sonarsource** [comment]
Hi @pmaieref,

Indeed, I think this would be better to discuss in [the Community](https://community.sonarsource.com/).
I'll still provide an answer here while the PR is open :-)

We do regular upgrades as they come. However, sometimes conflicting dependencies prevent us from updating faster.

You should be fine using older versions of TypeScript. We don't officially support newer versions. You might experience issues with new compiler options, which can be solved with a dedicated `tsconfig.sonar.json`. In other cases, we won't yet support new syntax in files that use it, but the rest of your analysis might work well. YMMV.

Keep in mind that if you are using SonarQube, you'll only get the updates when SonarQube itself is released and you update.
The SonarQube releases happen every ~2 months (latest version) or every ~18 months (long-term version).

Hope that helps!

**Comment by pmaieref** [comment]
@gabriel-vivas-sonarsource Highly appreciated the reply. If needed I'll create a post in the forum for any further discussions. Yes, we keep our SQ always on latest, thanks. :) 

Maybe you could discuss internally if a small update to https://docs.sonarsource.com/sonarqube/latest/analyzing-source-code/languages/javascript-typescript-css/ would make sense, i.e. adding some information regarding TS version support, tsconfig.sonar.json, etc.. 

Thank you very much for the quick reply here and good luck with the update! :)

**Comment by ilia-kebets-sonarsource** [comment]
@saberduck What are all the ruling diffs? Is there a general pattern?

File: its/ruling/src/test/expected/jsts/ace/typescript-S1874.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+"ace:demo/kitchen-sink/docs/typescript.ts": [
+35,
+46
```
Comment (ilia-kebets-sonarsource) [inline its/ruling/src/test/expected/jsts/ace/typescript-S1874.json:4]: these are weird:

```ts
module Sayings { // line 35
    export class Greeter {
        greeting: string;
        constructor (message: string) {
            this.greeting = message;
        }
        greet() {
            return "Hello, " + this.greeting;
        }
    }
}
module Mankala { // line 46
   export class Features {
       public turnContinues = false;
       public seedStoredCount = 0;
       public capturedCount = 0;
       public spaceCaptured = NoSpace;

       public clear() {
           this.turnContinues = false;
           this.seedStoredCount = 0;
           this.capturedCount = 0;
           this.spaceCaptured = NoSpace;
       }

       public toString() {
           var stringBuilder = "";
           if (this.turnContinues) {
               stringBuilder += " turn continues,";
           }
           stringBuilder += " stores " + this.seedStoredCount;
           if (this.capturedCount > 0) {
               stringBuilder += " captures " + this.capturedCount + " from space " + this.spaceCaptured;
           }
           return stringBuilder;
       }
   }
}
```

File: its/ruling/src/test/expected/jsts/ant-design/typescript-S4325.json [hunk 2]
```diff
@@ -8,8 +8,7 @@
 144
 ],
 "ant-design:components/anchor/Anchor.tsx": [
-27,
-258
```
Comment (ilia-kebets-sonarsource) [inline its/ruling/src/test/expected/jsts/ant-design/typescript-S4325.json:12]: this one is not great as well

File: its/ruling/src/test/expected/jsts/vuetify/typescript-S139.json [hunk 3]
```diff
@@ -41,9 +41,6 @@
 "vuetify:packages/vuetify/src/components/VDatePicker/VDatePicker.ts": [
 131
 ],
-"vuetify:packages/vuetify/src/components/VDatePicker/VDatePickerDateTable.ts": [
-48
-],
```
Comment (ilia-kebets-sonarsource) [inline its/ruling/src/test/expected/jsts/vuetify/typescript-S139.json:46]: This is really weird:

```ts
      return this.weekdayFormatter
        ? createRange(7).map(i => this.weekdayFormatter!(`2017-01-${first + i + 15}`)) // 2017-01-15 is Sunday <- this is line 48
        : createRange(7).map(i => ['S', 'M', 'T', 'W', 'T', 'F', 'S'][(i + first) % 7])
```

File: its/ruling/src/test/expected/jsts/ant-design/typescript-S4325.json (hunk 2 above)
Comment (saberduck) [reply its/ruling/src/test/expected/jsts/ant-design/typescript-S4325.json:12]: doesn't look incorrect https://github.com/microsoft/TypeScript/blob/main/src/lib/dom.generated.d.ts#L15864

File: its/ruling/src/test/expected/jsts/vuetify/typescript-S1438.json [hunk 4]
```diff
@@ -6949,54 +6949,6 @@
 535,
 537
 ],
-"vuetify:packages/vuetify/src/components/VDatePicker/VDatePickerDateTable.ts": [
-5,
```
Comment (ilia-kebets-sonarsource) [inline its/ruling/src/test/expected/jsts/vuetify/typescript-S1438.json:6953]: that's weird as well, is it the result of the TSconfig handling? all the missing semicolon issues disappeared on this file.

File: its/ruling/src/test/expected/jsts/vuetify/typescript-S2260.json [hunk 5]
```diff
@@ -0,0 +1,5 @@
+{
+"vuetify:packages/vuetify/src/components/VDatePicker/VDatePickerDateTable.ts": [
+null
+]
+}
```
Comment (ilia-kebets-sonarsource) [inline its/ruling/src/test/expected/jsts/vuetify/typescript-S2260.json:5]: What is null here?

**Comment by ilia-kebets-sonarsource** [comment]
It seems that all ruling issues disappeared from file `packages/vuetify/src/components/VDatePicker/VDatePickerDateTable.ts`

File: its/ruling/src/test/expected/jsts/vuetify/typescript-S1438.json (hunk 4 above)
Comment (ilia-kebets-sonarsource) [reply its/ruling/src/test/expected/jsts/vuetify/typescript-S1438.json:6953]: see https://github.com/SonarSource/SonarJS/pull/4816#issuecomment-2348993221

File: its/ruling/src/test/expected/jsts/vuetify/typescript-S139.json (hunk 3 above)
Comment (ilia-kebets-sonarsource) [reply its/ruling/src/test/expected/jsts/vuetify/typescript-S139.json:46]: see https://github.com/SonarSource/SonarJS/pull/4816#issuecomment-2348993221

File: its/ruling/src/test/expected/jsts/vuetify/typescript-S2260.json (hunk 5 above)
Comment (ilia-kebets-sonarsource) [reply its/ruling/src/test/expected/jsts/vuetify/typescript-S2260.json:5]: see https://github.com/SonarSource/SonarJS/pull/4816/#issuecomment-2348993221

**Review by ilia-kebets-sonarsource — APPROVED** [review]
There are changes in the ruling for these rules which depend on the TS type checker:
- [S4325: Redundant casts and non-null assertions should be avoided](https://sonarsource.github.io/rspec/#/rspec/S4325)
- [S1874: Code annotated as deprecated should not be used](https://sonarsource.github.io/rspec/#/rspec/S1874)

and weirdly one issue gone for rule [S6606: Nullish coalescing should be preferred](https://sonarsource.github.io/rspec/#/rspec/S6606/javascript) in [this location](https://github.com/SonarSource/SonarJS/pull/4816/files#diff-b169e621debe554d982c82f0ab53670018f4fc3fa9091d651e34cecb2bcd2926L19)

and one weirder thing: All issues are gone from the `vuetify:packages/vuetify/src/components/VDatePicker/VDatePickerDateTable.ts` file.

File: packages/ruling/tests/projects/vuetify.ruling.test.ts
Comment (ilia-kebets-sonarsource) [inline packages/ruling/tests/projects/vuetify.ruling.test.ts:1]: If you like the changes here, you should apply them to all the `packages/ruling/tests/projects/*.ruling.test.ts` files

**Comment by saberduck** [comment]
> It seems that all ruling issues disappeared from file `packages/vuetify/src/components/VDatePicker/VDatePickerDateTable.ts`

It seems that on Mac, we reached `Maximum call stack size exceeded` in file `vuetify:packages/vuetify/src/components/VDatePicker/VDatePickerDateTable.ts.` I had to pass `--stack-size=1000` for a ruling to pass on my Mac machine. On CI, the limit is probably higher (Linux). I will fix the CI in another PR so the failure is easier to debug.

**Pull request merged**